std/digest/xxh.d

```diff
@@ -0,0 +1,3352 @@
+/**
+ * Computes xxHash hashes of arbitrary data. xxHash hashes are either uint32_t, uint64_t or uint128_t quantities that are like a
+ * checksum or CRC, but are more robust and very performant.
+ *
+$(SCRIPT inhibitQuickIndex = 1;)
+
+$(DIVC quickindex,
+$(BOOKTABLE ,
+$(TR $(TH Category) $(TH Functions)
+)
+$(TR $(TDNW Template API) $(TD $(MYREF XXHTemplate)
+)
+)
+$(TR $(TDNW OOP API) $(TD $(MYREF XXH32Digest))
+)
+$(TR $(TDNW Helpers) $(TD $(MYREF xxh32Of))
+)
+)
+)
+
+ * This module conforms to the APIs defined in `std.digest`. To understand the
+ * differences between the template and the OOP API, see $(MREF std, digest).
+ *
+ * This module publicly imports $(MREF std, digest) and can be used as a stand-alone
+ * module.
+ *
+ * License:   $(HTTP www.boost.org/LICENSE_1_0.txt, Boost License 1.0).
+ *
+ * CTFE:
+ * Digests do not work in CTFE
+ *
+ * Authors:
+ * Carsten Schlote, Piotr Szturmaj, Kai Nacke, Johannes Pfau $(BR)
+ * The routines and algorithms are provided by the xxhash.[ch] source
+ * provided at $(I git@github.com:Cyan4973/xxHash.git).
+ *
+ * References:
+ *      $(LINK2 https://github.com/Cyan4973/xxHash, GitHub website of project)
+ *
+ * Source: $(PHOBOSSRC std/digest/xxh.d)
+ *
+ */
+
+/* xxh.d - A wrapper for the original C implementation */
+module std.digest.xxh;
+
+public import std.digest;
+
+///
+@safe unittest
+{
+    //Template API
+    import std.digest.md;
+
+    //Feeding data
+    ubyte[1024] data;
+    XXH_32 xxh;
+    xxh.start();
+    xxh.put(data[]);
+    xxh.start(); //Start again
+    xxh.put(data[]);
+    auto hash = xxh.finish();
+}
+
+///
+@safe unittest
+{
+    //OOP API
+    import std.digest.md;
+
+    auto xxh = new XXH32Digest();
+    ubyte[] hash = xxh.digest("abc");
+    assert(toHexString(hash) == "32D153FF", "Got " ~ toHexString(hash));
+
+    //Feeding data
+    ubyte[1024] data;
+    xxh.put(data[]);
+    xxh.reset(); //Start again
+    xxh.put(data[]);
+    hash = xxh.finish();
+}
+
+/* Port of C sources (release 0.8.1) to D language below */
+
+enum XXH_NO_STREAM = false;
+enum XXH_SIZE_OPT = 0;
+enum XXH_FORCE_ALIGN_CHECK = true;
+enum XXH32_ENDJMP = false;
+
+version (LittleEndian)
+    private immutable bool XXH_CPU_LITTLE_ENDIAN = true;
+else
+    private immutable bool XXH_CPU_LITTLE_ENDIAN = false;
+
+alias xxh_u8 = ubyte;
+alias xxh_u32 = uint;
+alias xxh_u64 = ulong;
+
+private uint32_t XXH_rotl32(uint32_t x, uint r) @trusted pure nothrow @nogc
+{
+    return (((x) << (r)) | ((x) >> (32 - (r))));
+}
+
+private uint64_t XXH_rotl64(uint64_t x, uint r) @trusted pure nothrow @nogc
+{
+    return (((x) << (r)) | ((x) >> (64 - (r))));
+}
```

Comment:
We should refactor this to avoid duplication with other digest implementations. The refactor should be done on another PR. See https://github.com/dlang/phobos/blob/93013129819529a2cdc23dc3f9bb9633dbe4631b/std/digest/murmurhash.d#L605 .

This is good for now, implementation-wise, but we should add compiler intrinsics for this @WalterBright . LLVM is smart enough to lower this tho https://godbolt.org/z/bj9fKj4eT

This can be `@safe` tho.

Reply:
Well, we have those already: https://dlang.org/phobos/core_bitop.html#.rol

Reply:
Thanks. Replaced code with rol() from core.bitop.